sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -320,6 +321,72 @@ func (n *DataSource) Progress() ProgressReportSnapshot {
 	return ProgressReportSnapshot{ID: n.SID.PtransformID, Name: n.Name, Count: c, pcol: pcol}
 }
 
+func (n *DataSource) getProcessContinuation() sdf.ProcessContinuation {
+	if u, ok := n.Out.(*ProcessSizedElementsAndRestrictions); ok {
+		return u.continuation
+	}
+	return nil
+}
+
+// Checkpoint attempts to split an SDF that has self-checkpointed (e.g. returned a
+// ProcessContinuation) and needs to be resumed later. If the underlying DoFn is not
+// splittable or has not returned a resuming continuation, the function returns an empty
+// SplitResult, a negative resumption time, and a false boolean to indicate that no split
+// occurred.
+func (n *DataSource) Checkpoint() (SplitResult, time.Duration, bool, error) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	pc := n.getProcessContinuation()
+	if pc == nil {
+		return SplitResult{}, -1 * time.Minute, false, nil
+	}
+	if !pc.ShouldResume() {
+		return SplitResult{}, -1 * time.Minute, false, nil
+	}
+
+	su := SplittableUnit(n.Out.(*ProcessSizedElementsAndRestrictions))
+
+	// Get the output watermark before splitting to avoid accidentally overestimating
+	ow := su.GetOutputWatermark()
```

Comment:
Oh nice - I was planning on following up in a PR to do this 😄 technically the comment doesn't apply anymore though since there's no risk of overestimating when checkpointing because execution has been fully paused at this point.

Reply:
Oh good, in that case I took the comment off. No need to add incorrect documentation here.